src/renderer/dx/DxFontRenderData.cpp

```diff
@@ -441,6 +446,77 @@ try
 }
 CATCH_RETURN()
 
+// Routine Description:
+// - Returns whether the user set or updated any of the font features to be applied
+bool DxFontRenderData::DidUserSetFeatures() const noexcept
+{
+    return _didUserSetFeatures;
+}
+
+// Routine Description:
+// - Clears our internal feature map and populates it with the list of standard features
+void DxFontRenderData::InitializeFeatureMap()
+{
+    _featureMap.clear();
+    _featureMap = {
+        { L"rlig", 1 },
+        { L"rclt", 1 },
+        { L"locl", 1 },
+        { L"ccmp", 1 },
+        { L"calt", 1 },
+        { L"liga", 1 },
+        { L"clig", 1 },
+        { L"kern", 1 },
+        { L"mark", 1 },
+        { L"mkmk", 1 },
+        { L"dist", 1 }
+    };
+}
+
+// Routine Description:
+// - Updates our internal map of font features with the given features
+// Arguments:
+// - features - the features to update our map with
+void DxFontRenderData::SetFeatures(std::unordered_map<std::wstring_view, uint32_t> features)
+{
+    // Populate our feature map with the standard list
+    InitializeFeatureMap();
+
+    // Update our feature map with the provided features
+    if (!features.empty())
+    {
+        for (const auto [tag, param] : features)
+        {
+            _featureMap[tag] = param;
+        }
+        _didUserSetFeatures = true;
+    }
+
+    // Convert the data to DWRITE_FONT_FEATURE and store it in a vector for CustomTextLayout
+    _featureVector.clear();
+    for (const auto [tag, param] : _featureMap)
+    {
+        const auto dwriteTag = DWRITE_MAKE_FONT_FEATURE_TAG(gsl::at(tag, 0), gsl::at(tag, 1), gsl::at(tag, 2), gsl::at(tag, 3));

```

Comment:
IIRC `til::at` is faster than `gsl::at` if we know that the array _does_ have that many elements. This isn't really gonna be a hot path, so this is really a nit